### ISSUE TITLE
change request: add a pure go implementation of a native package

### ISSUE THREAD

**Issue by dududko** [trinsic-id/okapi#177]
At the moment it is very difficult (but possible) to build and use the golang client because it depends on the `native` static library. However it is almost impossible to cross-compile the library for a different platform. It would help a lot if golang library would have a pure golang implementation. 

The `native` static library is build from the `native` package. From what I see the `native` package which is written in rust is only implementing some operations on protobuf structs, and those operations are not so complicated. Would it be possible to implement the same logic in golang, and use this pure golang solution in the golang api client ?

**Comment by tmarkovski** [comment]
@dududko While it is possible to add pure Go implementation, it would add another layer of code maintenance and development. With the FFI approach we currently use, all functionality is in written in Rust and made available to other languages using a common data model in Protobuf.
There's definitely some improvements we cam make in the Go package so that it'll use a precompiled static library in a more automated manner.

What was the biggest challenge you faced when using the Go client?

**Comment by dududko** [comment]
@tmarkovski I totally understand and support your decision to have a single static library and bindings for each language, especially due to the fact that it implements the cryptographic logic.

> What was the biggest challenge you faced when using the Go client?
The biggest challenge I faced was to cross-compile golang binary with c-go bindigs in transitive dependencies (your okapi/go library) from darwin or linux (host) to linux (target) and pack it in a docker image. I have managed to solve this task, but it would have been way easier of the head file and okapi static libraries for darwin/linux were published as release artifacts. I had to upload those files manually to our private s3 bucket to make it work.

Also we use bazel as a primarily build system in our company so I had to register additional toolchains, write new macro, create a patch for the `okapi` repo with missing `BUILD.bazel` files to be able to cross-compile cgo and prepare docker image. This took some time because it was our first encounter with cgo bindings. So yeah, it would have help a lot if you we also using bazel but I can only dream of this.
